torch/distributed/fsdp/_common_utils.py

```diff
@@ -509,7 +513,12 @@ def _no_dispatch_record_stream(tensor: torch.Tensor, stream: torch.Stream) -> No
     # FIXME record_stream doesn't work with non-cuda tensors
     if tensor.device.type not in ["cuda", torch._C._get_privateuse1_backend_name()]:
         return
-    with no_dispatch():
+
+    if not torch.distributed._functional_collectives.is_torchdynamo_compiling():
+        # Don't no dispatch under torch compile like this
+        with no_dispatch():
+            tensor.record_stream(stream)
+    else:
         tensor.record_stream(stream)
```

Comment:
is no_dispatch just a problematic graph-break under compile?  or something else?

what does compile do with tensor.record_stream anyway?

Reply:
The `no_dispatch` was added in https://github.com/pytorch/pytorch/pull/88014 cc @fegin 

Looking over the PR, it looks like this is because we don't actually support Stream arguments in torch dispatch, so it just chokes. If Dynamo is able to answer "are there any torch dispatch modes" active (it should answer False), a better version of this would just be to check if there are any modes before disabling dispatch.